pkg/cmd/server/server.go

```diff
@@ -195,7 +195,7 @@ func NewCommand(f client.Factory) *cobra.Command {
 	command.Flags().BoolVar(&config.restoreOnly, "restore-only", config.restoreOnly, "Run in a mode where only restores are allowed; backups, schedules, and garbage-collection are all disabled. DEPRECATED: this flag will be removed in v2.0. Use read-only backup storage locations instead.")
 	command.Flags().StringSliceVar(&config.disabledControllers, "disable-controllers", config.disabledControllers, fmt.Sprintf("List of controllers to disable on startup. Valid values are %s", strings.Join(controller.DisableableControllers, ",")))
 	command.Flags().StringSliceVar(&config.restoreResourcePriorities, "restore-resource-priorities", config.restoreResourcePriorities, "Desired order of resource restores; any resource not in the list will be restored alphabetically after the prioritized resources.")
-	command.Flags().StringVar(&config.defaultBackupLocation, "default-backup-storage-location", config.defaultBackupLocation, "Name of the default backup storage location.")
+	command.Flags().StringVar(&config.defaultBackupLocation, "default-backup-storage-location", config.defaultBackupLocation, "Name of the default backup storage location. DEPRECATED: this flag will be removed in v2.0. Use \"velero backup-location set --default\" instead.")
```

Comment:
What happens if both of these are set? We should document that behavior.

Reply:
Let me summarize the scenario I've thought:
1. During the upgrade, the BSL controller sets the defaulted BSL according to the BSL name.
2. The user could alert the defaulted BSL by velero CLI.
3. If the user changes the `velero server --default-backup-storage-location`, it takes no effect anymore.

The only way we can't prevent is the user directly edit BSL CR to have more than one BSLs having `.spec.default=true`.

Reply:
Let me summarize the scenario I've thought:
1. During the upgrade, the BSL controller sets the defaulted BSL according to the BSL name because none of the BSLs be marked as default.
2. The user could change the defaulted BSL by velero CLI.
3. Now that we have the default BSL. If the user changes the `velero server --default-backup-storage-location`, it takes no effect anymore. The only way to configure which BSL is the default is to change BSL CR (velero CLI _or_ `kubectl edit backupstoragelocations`).

The only way we can't prevent is the user directly edit BSL CRs to have more than one BSLs having `.spec.default=true`.

Reply:
That behavior sounds to me like it's not deprecating this flag, but making it entirely a no-op. To me, that's worse than removing the flag, because if it's present, users will expect it to work.

Deprecation means that it should still work, but is not the preferred method and will be removed.

Ideally, I think the flag should remain until complete removal and cause the server to execute similar logic to what the command does by altering the CRDs. This would happen at server startup, so it's a "one-time" operation, but the users can still use the new command to alter the default.

Reply:
> Ideally, I think the flag should remain until complete removal and cause the server to execute similar logic to what the command does by altering the CRDs. This would happen at server startup, so it's a "one-time" operation, but the users can still use the new command to alter the default.

sure, I'll make server-side and client-side change both work.

Reply:
I think another use case that might break the user's expectation, for example, we have 2 BSLs, one is `default` and the one is `secondary`.
1. During the upgrade, the BSL controller sets BSL `default` and default BSL.
2. The user changes the defaulted BSL to `secondary` by Velero CLI.
3. The Velero pod restart due to deleting the pod, then the velero server pod restart so it'll set the `default` BSL as the default.

Reply:
> It means that every time the velero pod start/restart, it might change the default BSL setting.

Yeah, that's a very good point. That's definitely not something we want, because that will send data to the wrong place based on other configuration changes. Perhaps the server-level option sets the default BSL once and only once, for migration purposes, then it doesn't work anymore?

How about we address these in a follow up PR to keep this one on track? I think the implementation here is pretty good, but the challenges for making it backwards compatible look trickier than I first thought and probably warrant their own set of focused changes.